### PR TITLE
Support top-level await

### LINKED ISSUE #5354: deno can await outside function, how to remove " error: await can only occur inside functions   "
deno can await outside function, how to remove " error: await can only occur inside functions   "

### PR DESCRIPTION
**Pull request by edemaine**
@GeoffreyBooth [mentioned](https://github.com/jashkenas/coffeescript/issues/5356#issuecomment-917525888) that CoffeeScript is missing support for [top-level await](https://v8.dev/features/top-level-await). This PR adds that support.

Mostly this involves doing fewer checks for invalidity.  The only hard part is when the code is wrapped in a top-level IFFE, where we need to detect whether that wrapper should be marked `async`.

Fixes #5354 (request from Deno side).

The REPL already supported top-level `await` by explicit wrapping in an `async function` closure. I think it's best to keep that code, because older versions of Node don't support top-level `await`.

### CODE REVIEW

File: test/async.coffee [hunk 1]
```diff
@@ -204,3 +204,21 @@ test "#3199: await multiline implicit object", ->
...
     eq undefined, y
+
+test "top-level await", ->
+  eq CoffeeScript.compile('await null'), '''
+    (async function() {
+      await null;
+
+    }).call(this);
+
+  '''
```
Comment (GeoffreyBooth) [inline test/async.coffee:215]: We have the `eqJS` helper that’s more concise, and does `compile` with `bare`:

```suggestion
  eqJS ‘await null’, ‘await null;’
```

File: test/async.coffee [hunk 2]
```diff
@@ -204,3 +204,21 @@ test "#3199: await multiline implicit object", ->
...
+  eq CoffeeScript.compile('-> await null'), '''
+    (function() {
+      (async function() {
+        return (await null);
+      });
+
+    }).call(this);
+
+  '''
```
Comment (GeoffreyBooth) [inline test/async.coffee:224]: I would change this to be a bit more realistic:

```suggestion
  eqJS 'do -> await null', '''
    (async function() {
      return (await null);
    })();
  '''
```

But it doesn’t belong in a “top level await” section, as this `await` isn’t top-level.

**Review by GeoffreyBooth — COMMENTED** [review]
Thank you for doing this! A few minor notes and then happy to land.

File: test/async.coffee (hunk 1 above)
Comment (edemaine) [reply test/async.coffee:215]: Ah, thanks, I figured you must have a helper like that.

I think there's value in testing both the bare and non-bare case, given that I had to write code to make the wrapper `async`.  Do you agree?  If so, should I include both tests like this, or make/modify helper to support nonbare?

File: test/async.coffee (hunk 2 above)
Comment (edemaine) [reply test/async.coffee:224]: What I meant to test here is that the wrapper is *not* marked `async` (in non-bare mode). I can give this test a better (and separate) name though.

File: test/async.coffee (hunk 1 above)
Comment (GeoffreyBooth) [reply test/async.coffee:215]: Let’s make separate tests for the non-bare wrapper, I think that’s worth calling out. I had no idea that’s what you were testing here.

I don’t think we need to modify the helper just for this one case. You can just use `CoffeeScript.compile` to test the wrapper.

**Comment by edemaine** [comment]
I've updated the tests to

* use `eqJS`
* extend `eqJS` to allow specifying compile options
* label tests better as to what's being tested

Happy to make further changes.

**Comment by edemaine** [comment]
Further revised the tests to check what they actually care about, which is a `startsWith` prefix.  `eqJS` is back to the original.

**Review by GeoffreyBooth — APPROVED** [review]

**Comment by GeoffreyBooth** [comment]
Thanks. I want to leave this open a little bit in case anyone else wants to review (@helixbass?) and then will land and cut a new release.

I added https://github.com/jashkenas/coffeescript/pull/5372 to add Node 16 support to CI but the tests fail there, so that should also probably be addressed before we land this. cc @helixbass on that too as it looks like it has to do with AST stuff.

**Comment by GeoffreyBooth** [comment]
Got CI fixed. Can you please rebase off of `master` or merge `master` into your branch? That should trigger a new test run that’ll include Node 16.

FYI besides rebuilding the regular compiler we also need to build the browser compiler (`cake build:browser`). We should add a CI step that rebuilds everything and then checks that there’s nothing different from the last git commit (which, if there was a difference, would mean that the author forgot to rebuild and commit one of the outputs).

**Pull request merged**